File: LC/python/Convert Sorted List to Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a list node
    # @return a tree node
    #soln1:convert to array
    #soln2:find mid and recursion to build bst
    def sortedListToBST(self, head):
        m = self.getLen(head)
        return self.buildBST(head,m,ListNode(0))
        
    def buildBST(self,head,cnt,res):
        if cnt == 0:
            res.next = head
            return None
        left = self.buildBST(head,cnt//2,res)
        root = TreeNode(res.next.val)
        res.next=res.next.next 
        right = self.buildBST(res.next,cnt-cnt//2-1,res)
        root.left = left
        root.right = right
        return root
        
    def getLen(self,head):
        size = 0
        while head:
            size+=1
            head = head.next
        return size

File: LC/java/Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    private ListNode curr;
    
    public TreeNode sortedListToBST(ListNode head) {
        int size;
        TreeNode root;
        
        size = getLen(head);
        curr = head;
        root = res(size);
        
        return root;
    }
    
    public int getLen(ListNode head){
        int size = 0;
            
        while(head!=null){
            size += 1;
            head = head.next;
        }
        return size;
    }
    
    public TreeNode res(int size){
        TreeNode root,left,right;
        
        if (size<=0){
            return null;
        }
        
        left = res(size/2);
        root = new TreeNode(curr.val);
        curr = curr.next;
        right = res(size - size/2 -1);
        
        root.left = left;
        root.right = right;
        return root;
    } 
}